### PR TITLE
Fix to UserSessionLimitsAuthenticator logic when client and realm limits exceeded. 

### LINKED ISSUE #46879: Exceeding Realm session cap assumes Client session cap is also exceeded
### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

Based on the logic in [User Sessions Limiter](https://github.com/keycloak/keycloak/blob/d600298df7612accc40c1a4c657275d7336ccd62/services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java#L88) if the realm session cap is exceeded then the client cap is assumed to be exceeded as well leading to this error 

```
2026-03-04 18:11:42,645 INFO  [org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator] (executor-thread-6521) Too many session in this realm for the current user. Session count: 104
2026-03-04 18:11:42,645 INFO  [org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator] (executor-thread-6521) Terminating oldest session
2026-03-04 18:11:42,645 INFO  [org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator] (executor-thread-6521) Logging out oldest -88 sessions
2026-03-04 18:11:42,646 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-6521) Uncaught server error: java.lang.IllegalArgumentException: -88
	at java.base/java.util.stream.ReferencePipeline.limit(ReferencePipeline.java:568)
	at org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator.logoutOldestSessions(UserSessionLimitsAuthenticator.java:206)
	at org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator.handleLimitExceeded(UserSessionLimitsAuthenticator.java:184)
	at org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator.authenticate(UserSessionLimitsAuthenticator.java:88)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:472)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:275)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:407)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:275)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:407)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:275)
	at org.keycloak.authentication.AuthenticationProcessor.authenticateOnly(AuthenticationProcessor.java:1118)
	at org.keycloak.protocol.oidc.grants.ResourceOwnerPasswordCredentialsGrantType.process(ResourceOwnerPasswordCredentialsGrantType.java:114)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:151)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint$quarkusrestinvoker$processGrantRequest_3903cccf0670c489ab77dc2ba1ba757573ec6d78.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:183)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:645)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

Since the logic assumes the targeted client for login is exceeding it's cap because the realm cap is exceeded a negative number is produced giving a 500. In reality, the targeted client only has 8/100 user sessions and another client has 92/100 user sessions.

### Version

26.5.3

### Regression

- [ ] The issue is a regression

### Expected behavior

No session created and 401 returned.

### Actual behavior

No session created and 500 returned.

### How to Reproduce?

Create two clients in the same realm, ClientA and ClientB. Set session limit for both and realm to 5. Login to ClientA three times. Login to ClientB three times. The third attempt to login to ClientB should give a 500.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by msdaly200**
Fix to UserSessionLimitsAuthenticator to correctly calculate no. of sessions to deny/terminate. 

Problem: Users were encountering an HTTP 500 Server Error (IllegalArgumentException) when their total realm sessions exceeded the configured limit, but their specific client sessions did not. When the realm limit was hit, the code logic evaluated the client sessions against the client limit to figure out how many sessions to kill.

Fix: 
[1] Code now calculates realmExcess and clientExcess independently and checks if (excess > 0) before attempting to terminate any sessions, so no negative numbers. 
[2] Checks client limits before the realm limits, If a user has too many sessions and a limit must be enforced, the system will sacrifice the user's oldest session from the exact app they are currently trying to access before it ever touches their sessions in other, unrelated applications. 
[3] Updated timestamp comparison logic as tests in new test framework were failing due to smaller time differences - reduce flakiness of migrated tests. 
[4] Tests for UserSessionLimitsAuthenticator were migrated to new test framework and split into smaller files. New tests were added to ensure edge cases were handled correctly. 

Closes #46879

### CODE REVIEW

File: tests/base/src/test/java/org/keycloak/tests/suites/Base2TestSuite.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
         "org.keycloak.tests.oid4vc",
         "org.keycloak.tests.securityprofile",
         "org.keycloak.tests.session",
+        "org.keycloak.tests.sessionlimits",
```
Comment (mabartos) [inline tests/base/src/test/java/org/keycloak/tests/suites/Base2TestSuite.java:27]: nitpick: Wouldn't it be better to have these session limits tests in the "org.keycloak.tests.session" package? As these tests might be relevant to every types of sessions + their characteristics (even the limits) WDYT?

File: services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java [hunk 2]
```diff
@@ -79,42 +82,45 @@ public void authenticate(AuthenticationFlowContext context) {
...
+            // Calculate excess only when new sessions would be created
+            if (newUserSession && userRealmLimit > 0 && userSessionCountForRealm >= userRealmLimit) {
+                realmExcess = userSessionCountForRealm - userRealmLimit + 1;
+            }
+            if (newClientSession && userClientLimit > 0 && userSessionCountForClient >= userClientLimit) {
+                clientExcess = userSessionCountForClient - userClientLimit + 1;
+            }
+
+            boolean realmLimitHandled = false;
+            boolean clientLimitHandled = false;
```
Comment (mabartos) [inline services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:94]: Is there any benefit of introducing the new counters for realmExcess + clientExcess?

File: services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java [hunk 3]
```diff
@@ -130,9 +136,9 @@ private int getIntConfigProperty(String key, Map<String, String> config) {
     private List<UserSessionModel> getUserSessionsForClientIfEnabled(List<UserSessionModel> userSessionsForRealm, ClientModel currentClient, int userClientLimit) {
         // Only count this users sessions for this client only in case a limit is configured, otherwise skip this costly operation.
         if (userClientLimit <= 0) {
+            logger.debugf("total user sessions for this keycloak client will not be counted. Will be logged as 0 (zero)");
```
Comment (mabartos) [inline services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:139]: +1

File: services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java [hunk 4]
```diff
@@ -192,31 +198,48 @@ private List<UserSessionModel> handleLimitExceeded(AuthenticationFlowContext con
...
         } else {
             logger.infof("Logging out oldest %s sessions", numberOfSessionsThatNeedToBeLoggedOut);
         }
 
+        // Find ID of the client the user is currently trying to log into
+        String currentClientId = context.getAuthenticationSession().getClient().getId();
+
+        // Prioritize current client's sessions for removal
+        Comparator<UserSessionModel> clientPriorityComparator = (s1, s2) -> {
+            boolean s1IsCurrent = s1.getAuthenticatedClientSessionByClient(currentClientId) != null;
```
Comment (mabartos) [inline services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:215]: Do we need to check if the user session is current for every comparison? AFAIK, this kind of check is done beforehand in the https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java#L75.

I'd just go with the comparator of the `getLastSessionRefresh` + if we need some more fine-grained tie-braker, go with the `UserSessionModel.getStarted()`(I see it's also in seconds, but might not be the same, mainly due to the `UserSessionEntity.mergeRemoteEntityWithLocalEntity()`. 

@msdaly200 WDYT?

**Review by mabartos — COMMENTED** [review]
@msdaly200 Thanks for PR! I've just added a few comments. 

Btw. the duplicated issue should be closed - as the issue should be resolved. If you'd like to do some improvements(like the comparator logic + tests), it'd be nice to create another issue(follow-up on the original issue) and reference it here. WDYT?

File: services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java (hunk 2 above)
Comment (msdaly200) [reply services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:94]: Good point, I can tighten that up more.

File: services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java (hunk 4 above)
Comment (msdaly200) [reply services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:215]: Good catch, will change comparator.

File: tests/base/src/test/java/org/keycloak/tests/suites/Base2TestSuite.java (hunk 1 above)
Comment (msdaly200) [reply tests/base/src/test/java/org/keycloak/tests/suites/Base2TestSuite.java:27]: This is the folder the migrate.sh script created.  As I broke down the large test into 3 smaller one, might be better to keep the folder name to reduce confusion. Also this folder tests only the UserSessionLimitsAuthenticator, other one is more session refresh and timeout tests.

**Comment by msdaly200** [comment]
Hi @mabartos so will create a new issue and add in those Authenticator changes and migrated tests, and close #46879 as a duplicate of #48040.

**Comment by msdaly200** [comment]
Closing PR as #46879 is a duplicate of #48040, which was already fixed by PR #48064 and merged yesterday. The fix is already in the main branch.  Have opened new task for the migrated tests - #48170
